src/rdf4cpp/rdf/writer/Serialization.hpp

```diff
@@ -0,0 +1,178 @@
+#ifndef RDF4CPP_SERIALIZATION_HPP
+#define RDF4CPP_SERIALIZATION_HPP
+
+#include <functional>
+
+#include <rdf4cpp/rdf/Literal.hpp>
+#include <rdf4cpp/rdf/Quad.hpp>
+#include <rdf4cpp/rdf/writer/BufWriter.hpp>
+#include <rdf4cpp/rdf/writer/SerializationState.hpp>
+
+namespace rdf4cpp::rdf::writer {
+template<OutputFormat F>
+concept format_has_graph = (F == OutputFormat::NQuads || F == OutputFormat::TriG);
+
+template<OutputFormat F>
+concept format_has_prefix = (F == OutputFormat::Turtle || F == OutputFormat::TriG);
+
+struct TypeIRIPrefix {
+    std::string_view prefix;
+    std::string_view shorthand;
+};
+static constexpr std::array iri_prefixes = {
+        TypeIRIPrefix{"http://www.w3.org/2001/XMLSchema#", "xsd"},
+        TypeIRIPrefix{"http://www.w3.org/1999/02/22-rdf-syntax-ns#", "rdf"},
+};
+
+bool write_prefix_data(void *buffer, Cursor &cursor, FlushFunc flush);
+template<OutputFormat F>
+bool write_prefix(void *const buffer, Cursor &cursor, FlushFunc const flush) {
+    if constexpr (format_has_prefix<F>) {
+        return write_prefix_data(buffer, cursor, flush);
+    }
+    return true;
+}
+template<OutputFormat F>
+bool flush_state(void *const buffer, Cursor &cursor, FlushFunc const flush, SerializationState *state) {
+    if constexpr (format_has_prefix<F>) {
+        if (state == nullptr)
+            return false;
+        if (!state->active_predicate.null() || !state->active_subject.null())
+            if (!write_str(" .\n", buffer, &cursor, flush))
+                return false;
+        if (!state->active_graph.null())
+            if (!write_str("}\n", buffer, &cursor, flush))
+                return false;
+        state->active_predicate = Node::make_null();
+        state->active_subject = Node::make_null();
+        state->active_graph = Node::make_null();
+    }
+    return true;
+}
+
+template<OutputFormat F>
+bool serialize(const Quad &s, void *const buffer, Cursor &cursor, FlushFunc const flush, SerializationState *state) {
+    auto write_node = &Node::serialize_short_form;
```

Comment:
Use static dispatch instead of dynamic. I.e.

```c++
struct ShortFormNodeSerializer {
    static bool serialize(Node term, void *const buffer, Cursor *cursor, FlushFunc const flush) noexcept {
        return term.serialize_short_form(buffer, cursor, flush);
    }
};

struct LongFormNodeSerializer {
    static bool serialize(Node term, void *const buffer, Cursor *cursor, FlushFunc const flush) noexcept {
        return term.serialize(buffer, cursor, flush);
    }
};

[..]

using NodeSer = std::conditional_t<format_has_prefix<F>, ShortFormNodeSerializer, LongFormNodeSerializer>;

NodeSer::serialize(...) // this was write_node(...) before.

```